crates/core/src/binding.rs

```diff
@@ -324,6 +325,39 @@ pub(crate) fn linearize_binding<'a>(
 }
 
 impl<'a> Binding<'a> {
+    pub(crate) fn from_node(node: NodeWithSource<'a>) -> Self {
+        Self::Node(node.source, node.node)
+    }
+
+    /// Returns the node this binding applies to.
+    ///
+    /// Unlike [`Self::get_node()`], this will return the exact child node if
+    /// the binding applies to a list.
+    pub(crate) fn as_node(&self) -> Option<NodeWithSource<'a>> {
+        match self {
+            Self::Node(src, node) => Some(NodeWithSource::new(node.clone(), src)),
+            Self::List(src, node, field) => node
+                .child_by_field_id(*field)
+                .map(|child| NodeWithSource::new(child, src)),
+            Self::Empty(..) | Self::String(..) | Self::ConstantRef(..) | Binding::FileName(..) => {
+                None
+            }
+        }
+    }
+
+    /// Returns the node from which this binding was created.
+    ///
+    /// This differs from [`Self::as_node()`] in that it returns the node for an
+    /// entire list when the binding applies to a list item.
+    pub(crate) fn get_node(&self) -> Option<NodeWithSource<'a>> {
```

Comment:
I made these two separate implementations to mirror the original implementations. But I notice if I change all calls to `get_node()` with `as_node()`, the tests also remain green... and the other way around as well. It makes me wonder whether this is simply a case of lacking test coverage, or was this divergence in the implementations an oversight in the original implementation?

Reply:
Likely an oversight, we can consolidate on get_node.

Reply:
I would consolidate using get_node, or maybe even return None for lists/empty.  rather than as_node. as_node looks buggy to me for two reasons. `child_by_field_id` gives the first node in the field, so as_node does not distinguish between fields with multiple nodes or single nodes. additionally child_by_field_id (if it behaves like children_by_field_id) includes unnamed nodes which we do not care about while matching